cmd/plugin/builder/plugin/build-packages.go

```diff
@@ -50,50 +53,80 @@ func (bpo *BuildPluginPackageOptions) BuildPluginPackages() error {
 
 	log.Infof("Using plugin binary artifacts from %q", bpo.BinaryArtifactDir)
 
-	for i := range pluginManifest.Plugins {
-		for _, osArch := range cli.AllOSArch {
-			for _, version := range pluginManifest.Plugins[i].Versions {
-				pluginBinaryFilePath := filepath.Join(bpo.BinaryArtifactDir, osArch.OS(), osArch.Arch(),
-					pluginManifest.Plugins[i].Target, pluginManifest.Plugins[i].Name, version,
-					cli.MakeArtifactName(pluginManifest.Plugins[i].Name, osArch))
+	// Limit the number of concurrent operations we perform so we don't overwhelm the system.
+	maxConcurrent := helpers.GetMaxParallelism()
+	guard := make(chan struct{}, maxConcurrent)
+	var wg sync.WaitGroup
+	// Mix up IDs so we don't always get the same set.
+	randSkew := rand.Intn(len(helpers.Identifiers)) // nolint:gosec
+	errorList := []error{}
+
+	generatePluginPackage := func(p cli.Plugin, osArch cli.Arch, version, id string) {
+		defer func() {
+			<-guard
+			wg.Done()
+		}()
+
+		pluginBinaryFilePath := filepath.Join(bpo.BinaryArtifactDir, osArch.OS(), osArch.Arch(),
+			p.Target, p.Name, version, cli.MakeArtifactName(p.Name, osArch))
+
+		if !utils.PathExists(pluginBinaryFilePath) {
+			return
+		}
 
-				if !utils.PathExists(pluginBinaryFilePath) {
-					continue
-				}
+		// Check whether the binary exec file is valid
+		valid, err := helpers.ValidatePluginBinary(pluginBinaryFilePath)
```

Comment:
This function call will be the first output shown to the user(`Validating Plugin Binary file: <path>`.
This output does not have the special thread ID, but it does show the path of the plugin binary, so the user can figure things out.

I suggest that we add a printout at the very beginning of each execution of the new `generatePluginPackage()` to indicate we are starting a new thread.

As for `helpers.ValidatePluginBinary()` I leave it up to you to decide if you want to include the threadId in the output or not.  Or even remove the printout completely (it is not that important).  Actually, I'm thinking removing it is maybe better?

Reply:
Good point. Updated to pass threadID to the `ValidatePluginBinary` to make logs consistent.